### PR TITLE
Added function to calculate cumulative GPA(including unfinished semesters)

### LINKED ISSUE #1: Show the effect of current semester to cumulative GPA
Add the ability to show what the cumulative GPA will be once the current semester ends with the current GPA.

This is will require some way to figure out how many semesters the student has been at SAS for.

### PR DESCRIPTION
**Pull request by insertcustomname**
Closes #1 

Calculate cumulative GPA and added optional parameter creditHour to Course class.

Signed-off-by: Suhas Hariharan <hariharan774531@sas.edu.sg>

### CODE REVIEW

**Comment by insertcustomname** [comment]
Sorry I messed up the other branch so I just created a new pull request/branch with it all in one commit.

**Comment by insertcustomname** [comment]
I just added something so it should calculate cumulative including current semester, I tried using the begdate thing to figure out whether the semester was over or not.

File: src/js/saspowerschoolff.js [hunk 1]
```diff
@@ -106,7 +117,7 @@ function main_page () {
                 if (gradeToGPA($first_grade.text()) !== -1) {
                     new (Vue.extend(ClassGrade))({
                         propsData: {
-                            course: new Course("", `https://powerschool.sas.edu.sg/guardian/${$first_grade.attr('href')}`, $first_grade.text()),
+                            course: new Course("", `https://powerschool.sas.edu.sg/guardian/${$first_grade.attr('href')}`, $first_grade.text(), 0, ""),
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]: Why is this change necessary?

File: src/js/saspowerschoolff.js [hunk 2]
```diff
@@ -70,6 +72,15 @@ function analytics_message (action_input) {
...
 }
+function calculate_credit_hours (text) {
+    if (text.substr(0, 2) === "IS") {
+        return 0.25;
+    } else if (text.includes("A-B")) {
+        return 1;
+    } else {
+        return 0.5;
+    }
+}
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]: Let's reuse the function we already have, DRY and all.

File: src/js/saspowerschoolff.js [hunk 3]
```diff
@@ -117,7 +128,9 @@ function main_page () {
         }
         if ($course.length === 1) {
             const temp = $course.parents().eq(1).children("td[align=left]").text().match(".*(?=Details)")[0];
-            courses.push(new Course(temp.trim(), `https://powerschool.sas.edu.sg/guardian/${$course.attr('href')}`, $course.text()));
+            let creditHour = 0;
+            creditHour = calculate_credit_hours($($course.parents().get(1)).find("td").get(0).textContent);
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]:
```suggestion
            const creditHour = calculate_credit_hours($($course.parents().get(1)).find("td").get(0).textContent);
```

File: src/js/saspowerschoolff.js (hunk 2 above)
Comment (insertcustomname) [reply src/js/saspowerschoolff.js]: Sure, didn't realise there was already something for that, I'll modify it.

File: src/js/saspowerschoolff.js (hunk 1 above)
Comment (insertcustomname) [reply src/js/saspowerschoolff.js]: That's my bad, I originally had that also including credit hours and than I realised it wasn't necessary for credit hours to be included there so I deleted the creditHour parameter but left the rest. I'll fix that

**Comment by insertcustomname** [comment]
Is it okay with you if I take total_add out of the calculate_gpa function and rename it calcualte_credit_hours?

**Comment by gary-kim** [comment]
> Is it okay with you if I take total_add out of the calculate_gpa function and rename it calcualte_credit_hours?

Yes. You don't have to ask :).

File: src/js/models/Course.js [hunk 4]
```diff
@@ -37,19 +38,25 @@ export default class Course {
...
         this.#assignments = assignments;
+        this.#creditHour = creditHour;
     }
 
     get name () {
         return this.#name;
     }
 
+    get creditHour () {
+        return this.#creditHour;
```
Comment (gary-kim) [inline src/js/models/Course.js]: Just had an idea. Could we make it so that `get creditHour` uses the name of the course to calculate it rather than it having to be provided?

We could do this during the cleanup pr or now.

Comment (insertcustomname) [reply src/js/models/Course.js]: That's a great idea, I'll add that and refactor to use that instead.

**Comment by insertcustomname** [comment]
I made the changes, does this look ready to merge?

File: src/js/models/Course.js (hunk 4 above)
Comment (gary-kim) [reply src/js/models/Course.js]: Wasn't the way I was thinking but that works too!

Comment (insertcustomname) [reply src/js/models/Course.js]: Do you think we can merge it now or is there anything else you think needs to be modified?

File: src/js/saspowerschoolff.js [hunk 5]
```diff
@@ -188,6 +203,85 @@ function login_page () {
...
+    let element_list = [];
+    const fetches = [];
+    // Fetches grade history page
+    const gpas = fetch("https://powerschool.sas.edu.sg/guardian/termgrades.html")
+        .then(response => response.text())
+        .then(data => {
+            const el = document.createElement("html");
+            el.innerHTML = data;
+            const tabs = el.getElementsByClassName("tabs")[0].getElementsByTagName("li");
+            for (let i = 0; i < tabs.length; i++) {
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]:
```suggestion
            // Iterate until the end of tabs or until no longer at a high school semester.
            for (let i = 0; i < tabs.length && /HS$/.test(tabs.innerText); i++) {
```
To make sure that middle school grades are not included.

File: src/js/saspowerschoolff.js [hunk 6]
```diff
@@ -162,6 +171,12 @@ function main_page () {
         },
     }).$mount(".hypo-grade-div-fixed");
 }
+function show_cumulative_gpa (courses) {
+    $("#calculateCumulative").hide();
+    calculate_cumulative_gpa(courses).then(cumulative_gpa => {
+        $("table[border='0'][cellpadding='3'][cellspacing='1'][width='100%']").prepend(`<tr><td align="center">Cumulative GPA: ${cumulative_gpa.toFixed(2)}</td></tr>`);
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]:
```suggestion
        $("table[border='0'][cellpadding='3'][cellspacing='1'][width='100%']").prepend(`<tr><td align="center">Cumulative GPA (9-12): ${cumulative_gpa.toFixed(2)}</td></tr>`);
```
I think this makes it a little clearer. Just a suggestion though, if you think this is clear enough as is, just keep it as is.

File: src/js/saspowerschoolff.js [hunk 7]
```diff
@@ -117,7 +120,7 @@ function main_page () {
         }
         if ($course.length === 1) {
             const temp = $course.parents().eq(1).children("td[align=left]").text().match(".*(?=Details)")[0];
-            courses.push(new Course(temp.trim(), `https://powerschool.sas.edu.sg/guardian/${$course.attr('href')}`, $course.text()));
+            courses.push(new Course(temp.trim(), `https://powerschool.sas.edu.sg/guardian/${$course.attr('href')}`, $course.text(), 0, ""));
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]:
```suggestion
            courses.push(new Course(temp.trim(), `https://powerschool.sas.edu.sg/guardian/${$course.attr('href')}`, $course.text()));
```

File: src/js/saspowerschoolff.js (hunk 5 above)
Comment (insertcustomname) [reply src/js/saspowerschoolff.js]: Oh didn't think of that, thanks.

**Comment by insertcustomname** [comment]
Not quite sure what happened, I just commited your suggestions than ran git rebase HEAD~7 --signoff to add signoffs as per what the DCO prompted me to do, now it seems that theres conflicts?

**Comment by insertcustomname** [comment]
Could I just do git reset --hard 25edb97 to go back to before?

**Comment by gary-kim** [comment]
You shouldn't need to. You can just run `git fetch && git rebase origin/master` and that will fix the conflicts.

**Comment by insertcustomname** [comment]
Okay, it seems to have added a bunch more files to the files changed as well for this commit, would that all be resolved by rebasing?

File: src/js/saspowerschoolff.js (hunk 5 above)
Comment (insertcustomname) [reply src/js/saspowerschoolff.js]: Does this work for you? I just tested this and it results in my cumulative being calculate incorrectly.

Comment (insertcustomname) [reply src/js/saspowerschoolff.js]: Just needed to change it to tabs[i].innerText.

**Comment by insertcustomname** [comment]
Should be ready to merge?

File: src/js/saspowerschoolff.js [hunk 8]
```diff
@@ -188,6 +203,85 @@ function login_page () {
...
+                                }
+                            }
+                            all_courses.push(courses);
+                        }));
+            }
+            // Calculates cumulative GPA based on credit hours per semester and gpa for each semester.
+            let include_current_semester = false;
+            if (current_courses.length !== 0) {
+                for (let i = 0; i < current_courses.length; i++) {
+                    if (current_courses[i].link.includes("begdate")) {
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]: The better way to do this would be something like
```suggestion
                    if (new URL(current_courses[i].link).searchParams.get("begdate"))  {
```
but we can deal with that during the cleanup if you'd like.

File: src/js/saspowerschoolff.js [hunk 9]
```diff
@@ -188,6 +203,85 @@ function login_page () {
...
+                                    all_courses.push(courses);
+                                    courses = [];
+                                }
+                                if (element_list[t].getElementsByTagName("th").length > 0) {
+                                    continue;
+                                } else {
+                                    const $prev_course = element_list[t];
+                                    // Creates course object with each course from grade history page
+                                    const course = new Course($prev_course.getElementsByTagName("td")[0].textContent.trim(), "",
+                                        $prev_course.getElementsByTagName("td")[1].textContent.trim(), 0, "");
```
Comment (gary-kim) [inline src/js/saspowerschoolff.js]:
```suggestion
                                        $prev_course.getElementsByTagName("td")[1].textContent.trim());
```

File: src/js/models/Course.js (hunk 4 above)
Comment (gary-kim) [reply src/js/models/Course.js]: For this, I was thinking more along the lines of changing the getter and getting rid of the `#creditHour` variable but maybe this could come in useful in the future to overwrite the automatically calculated credit hour so it could be useful.

File: src/js/saspowerschoolff.js (hunk 8 above)
Comment (insertcustomname) [reply src/js/saspowerschoolff.js]: Sure, thanks okay if I merge it now?

Comment (gary-kim) [reply src/js/saspowerschoolff.js]: If the PR's approved by someone with write access, you don't need to ask :)
